protocol_rfcs/type-widening.md

```diff
@@ -0,0 +1,143 @@
+`fromType`| required | The type of the column or field before the type change.
+`toType`| required | The type of the column or field after the type change.
+`fieldPath`| optional | When updating the type of a map key/value or array element only: the path from the struct field holding the metadata to the map key/value or array element that was updated.
+
+The `fieldPath` value is "key", "value" and "element"  when updating resp. the type of a map key, map value and array element.
+The `fieldPath` value for nested maps and nested arrays are prefixed by their parents's path, separated by dots.
+
+The following is an example for the definition of a column that went through two type changes:
+```json
+{
+    "name" : "e",
+    "type" : "long",
+    "nullable" : true,
+    "metadata" : { 
+      "delta.typeChanges": [
+        {
+          "tableVersion": 1,
+          "fromType": "short",
+          "toType": "int"
+        },
+        {
+          "tableVersion": 5,
+          "fromType": "int",
+          "toType": "long"
+        }
+      ]
+    }
+  }
+```
+
+The following is an example for the definition of a column after changing the type of a map key:
+```json
+{
+    "name" : "e",
+    "type" : {
+      "type": "map",
+      "keyType": "double",
+      "valueType": "int",
+      "valueContainsNull": true
+    },
+    "nullable" : true,
+    "metadata" : { 
+      "delta.typeChanges": [
+        {
+          "tableVersion": 2,
+          "fromType": "float",
+          "toType": "double",
+          "fieldPath": "key"
+        }
+      ]
+    }
+  }
+```
+
+The following is an example for the definition of a column after changing the type of a map value nested in an array:
+```json
+{
+    "name" : "e",
+    "type" : {
+      "type": "array",
+      "elementType": {
+        "type": "map",
+        "keyType": "string",
+        "valueType": "decimal(10, 4)",
+        "valueContainsNull": true
+      },
+      "containsNull": true
+    },
+    "nullable" : true,
+    "metadata" : { 
+      "delta.typeChanges": [
+        {
+          "tableVersion": 2,
+          "fromType": "decimal(6, 2)",
+          "toType": "decimal(10, 4)",
+          "fieldPath": "element.key"
+        }
+      ]
+    }
+  }
+```
+
+## Writer Requirements for Type Widening
+
+When Type Widening is enabled (when the table property `delta.enableTypeWidening` is set to `true`), then:
+- Writers should allow updating the table schema to apply an **allowed type change** to a column, struct field, map key/value or array element.
+- Writers must record type change information in the `metadata` of the nearest ancestor [StructField](#struct-field). See [Type Change Metadata](#type-change-metadata).
+
+When Type Widening is supported (when the `writerFeatures` field of a table's `protocol` action contains `enableTypeWidening`), then:
+- Writers must set the `defaultRowCommitVersion` field in new `add` actions to the version number of the log enty containing the `add` action.
```

Comment:
They must also _preserve_ this field when the `add` action is replicated, including in checkpoints and when repeating an `add` action to updating metadata.

Reply:
Added.